server/src/tabbing/tabCalculator.ts

```diff
@@ -1,152 +1,81 @@
-// TODO: improve this to be smoother - only get local maxima which are bigger than multiple prior values
-function getLocalMaximaIndices(values: number[]): number[] {
-    // handle edge cases
-    if (values.length == 0) return [];
-    if (values.length == 1) return [0];
-
-    const indices: number[] = [];
-
-    // handle first index being a local maximum
-    if (values[0] >= values[1]) {
-        indices.push(0);
-    }
-    // handle middle values being a local maximum
-    for (var i = 1; i < values.length - 1; ++i) {
-        // use >= for next value in order to get only the first of equal values
-        if (values[i] > values[i - 1] && values[i] >= values[i + 1]) {
-            indices.push(i);
-        }
-    }
-    // handle last index being a local maximum
-    if (values[values.length-1] > values[values.length-2]) {
-        indices.push(values.length-1);
-    }
-
-    return indices;
-}
-
-// For each value in the input array, returns a pair.
-// The first element in the pair is the number of prior values which
-// the element is larger than or equal to.
-// The second element in the pair is the number of subsequent values which
-// the element is larger than.
-// e.g. with input array [1, 2, 3, 4, 3, 2, 3, 5], the output array is
-// [(0, 0), (1, 0), (2, 0), (3, 3), (0, 1), (0, 0), (2, 0), (7, 0)]
-// This allows us to have a single function that will work for whatever
-// maximum values we need - e.g. if we want to only look at values which
-// are the local maximums of an 8-sample region, or which must be larger
-// than the 3 prior samples and the 6 subsequent samples, and so forth.
-// This implementation is O(n).
-function smoothLocalMaxima(values: number[]): SmoothMaxima {
-    const previousValues: number[] = new Array(values.length);
-    const nextValues: number[] = new Array(values.length);
-    const stack: number[] = [];
 
-    for (var i = 0; i < values.length; ++i) {
-        while (stack.length > 0 && values[stack[stack.length-1]] <= values[i]) {
-            const idx = stack.pop();
-            nextValues[idx] = i - idx - 1;
+    // note is a MIDI-tuning-standard note index, between 0 (C_{-1}) and 127 (G_9).
+    // Low E on guitar is E_2 (80 Hz, index 40).
+    // For now, we take the trivial idea that the note is played on the highest-pitched
+    // string that can play it.
+    // Standard tuning: (string index, note, note index)
+    // 0 E_4 64
+    // 1 B_3 59
+    // 2 G_3 55
+    // 3 D_3 50
+    // 4 A_2 45
+    // 5 E_2 40
+    //
+    // see en.wikipedia.org/wiki/Scientific_pitch_notation#Table_of_note_frequencies
+    // and en.wikipedia.org/wiki/Guitar_tunings#Standard
+    public static getStringAndFret(note: number, tuning: Tuning): StringFret {
+        const notes: number[] = tuning.getNotes();
+        for (var i = 0; i < notes.length; ++i) {
+            if (note >= notes[i]) {
+                return { stringIdx: i, fret: note - notes[i] };
```

Comment:
I mean, there are 24-fret guitars. I'm not too concerned about that issue, because the odds of crepe incorrectly picking up a frequency that high is pretty slim. Worst-case we tell someone to play the 32nd fret on the high e :P